Subject: [PATCH] staging: fclk: Enable pl clock with sysfs entry for set_rate
---
Index: Documentation/devicetree/bindings/clock/xlnx,fclk.yaml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Documentation/devicetree/bindings/clock/xlnx,fclk.yaml b/Documentation/devicetree/bindings/clock/xlnx,fclk.yaml
new file mode 100644
--- /dev/null	(revision 2aad4db65b68532992f8ed292ffd10c3a3a12bd6)
+++ b/Documentation/devicetree/bindings/clock/xlnx,fclk.yaml	(revision 2aad4db65b68532992f8ed292ffd10c3a3a12bd6)
@@ -0,0 +1,33 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/clock/xlnx,fclk.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Xilinx PL clocks
+
+maintainers:
+  - Shubhrajyoti Datta <shubhrajyoti.datta@amd.com>
+
+description:
+  The clock consumer driver used for handle clocks between PS/PL but also
+  provide a way to control clock from System Controller to DUT.
+
+properties:
+  compatible:
+    const: xlnx,fclk
+
+  clocks:
+    maxItems: 1
+
+required:
+  - clocks
+
+additionalProperties: false
+
+examples:
+  -
+    clk {
+      compatible = "xlnx,fclk";
+      clocks = <&clkc 71>;
+    };
Index: drivers/staging/Kconfig
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/drivers/staging/Kconfig b/drivers/staging/Kconfig
--- a/drivers/staging/Kconfig	(revision 519d8ae2f3ef6ed7c1af010cb43c4dbfd3207a8d)
+++ b/drivers/staging/Kconfig	(revision 2aad4db65b68532992f8ed292ffd10c3a3a12bd6)
@@ -64,4 +64,6 @@

 source "drivers/staging/vme_user/Kconfig"

+source "drivers/staging/fclk/Kconfig"
+
 endif # STAGING
Index: drivers/staging/Makefile
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/drivers/staging/Makefile b/drivers/staging/Makefile
--- a/drivers/staging/Makefile	(revision 519d8ae2f3ef6ed7c1af010cb43c4dbfd3207a8d)
+++ b/drivers/staging/Makefile	(revision 2aad4db65b68532992f8ed292ffd10c3a3a12bd6)
@@ -21,3 +21,4 @@
 obj-$(CONFIG_BCM2835_VCHIQ)	+= vc04_services/
 obj-$(CONFIG_XIL_AXIS_FIFO)	+= axis-fifo/
 obj-$(CONFIG_FIELDBUS_DEV)     += fieldbus/
+obj-$(CONFIG_XILINX_FCLK)	+= fclk/
Index: drivers/staging/fclk/Kconfig
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/drivers/staging/fclk/Kconfig b/drivers/staging/fclk/Kconfig
new file mode 100644
--- /dev/null	(revision 2aad4db65b68532992f8ed292ffd10c3a3a12bd6)
+++ b/drivers/staging/fclk/Kconfig	(revision 2aad4db65b68532992f8ed292ffd10c3a3a12bd6)
@@ -0,0 +1,9 @@
+#
+# Xilinx PL clk enabler
+#
+
+config XILINX_FCLK
+	tristate "Xilinx PL clock enabler"
+	depends on COMMON_CLK && OF
+	help
+	  Support for the Xilinx fclk clock enabler.
Index: drivers/staging/fclk/Makefile
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/drivers/staging/fclk/Makefile b/drivers/staging/fclk/Makefile
new file mode 100644
--- /dev/null	(revision 2aad4db65b68532992f8ed292ffd10c3a3a12bd6)
+++ b/drivers/staging/fclk/Makefile	(revision 2aad4db65b68532992f8ed292ffd10c3a3a12bd6)
@@ -0,0 +1,1 @@
+obj-$(CONFIG_XILINX_FCLK)	+= xilinx_fclk.o
Index: drivers/staging/fclk/TODO
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/drivers/staging/fclk/TODO b/drivers/staging/fclk/TODO
new file mode 100644
--- /dev/null	(revision 2aad4db65b68532992f8ed292ffd10c3a3a12bd6)
+++ b/drivers/staging/fclk/TODO	(revision 2aad4db65b68532992f8ed292ffd10c3a3a12bd6)
@@ -0,0 +1,2 @@
+TODO:
+	- Remove this hack and clock adapt all the drivers.
Index: drivers/staging/fclk/xilinx_fclk.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/drivers/staging/fclk/xilinx_fclk.c b/drivers/staging/fclk/xilinx_fclk.c
new file mode 100644
--- /dev/null	(revision 2aad4db65b68532992f8ed292ffd10c3a3a12bd6)
+++ b/drivers/staging/fclk/xilinx_fclk.c	(revision 2aad4db65b68532992f8ed292ffd10c3a3a12bd6)
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Xilinx fclk clock driver.
+ * Copyright (c) 2017 - 2020 Xilinx Inc.
+ */
+
+#include <linux/clk.h>
+#include <linux/clk-provider.h>
+#include <linux/errno.h>
+#include <linux/io.h>
+#include <linux/module.h>
+#include <linux/of_device.h>
+#include <linux/platform_device.h>
+
+struct fclk_state {
+	struct device	*dev;
+	struct clk	*pl;
+};
+
+/* Match table for of_platform binding */
+static const struct of_device_id fclk_of_match[] = {
+	{ .compatible = "xlnx,fclk",},
+	{ /* end of list */ },
+};
+MODULE_DEVICE_TABLE(of, fclk_of_match);
+
+static ssize_t set_rate_show(struct device *dev,
+			     struct device_attribute *attr, char *buf)
+{
+	struct fclk_state *st = dev_get_drvdata(dev);
+
+	return scnprintf(buf, PAGE_SIZE, "%lu\n", clk_get_rate(st->pl));
+}
+
+static ssize_t set_rate_store(struct device *dev,
+			      struct device_attribute *attr,
+			      const char *buf, size_t count)
+{
+	int ret = 0;
+	unsigned long rate;
+	struct fclk_state *st = dev_get_drvdata(dev);
+
+	ret = kstrtoul(buf, 0, &rate);
+	if (ret)
+		return -EINVAL;
+
+	rate = clk_round_rate(st->pl, rate);
+	ret = clk_set_rate(st->pl, rate);
+
+	return ret ? ret : count;
+}
+
+static DEVICE_ATTR_RW(set_rate);
+
+static const struct attribute *fclk_ctrl_attrs[] = {
+	&dev_attr_set_rate.attr,
+	NULL,
+};
+
+static const struct attribute_group fclk_ctrl_attr_grp = {
+	.attrs = (struct attribute **)fclk_ctrl_attrs,
+};
+
+static int fclk_probe(struct platform_device *pdev)
+{
+	struct fclk_state *st;
+	int ret;
+	struct device *dev = &pdev->dev;
+
+	st = devm_kzalloc(&pdev->dev, sizeof(*st), GFP_KERNEL);
+	if (!st)
+		return -ENOMEM;
+
+	st->dev = dev;
+	platform_set_drvdata(pdev, st);
+
+	st->pl = devm_clk_get(&pdev->dev, NULL);
+	if (IS_ERR(st->pl))
+		return PTR_ERR(st->pl);
+
+	ret = clk_prepare_enable(st->pl);
+	if (ret) {
+		dev_err(&pdev->dev, "Unable to enable clock.\n");
+		return ret;
+	}
+
+	ret = sysfs_create_group(&dev->kobj, &fclk_ctrl_attr_grp);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+static void fclk_remove(struct platform_device *pdev)
+{
+	struct fclk_state *st = platform_get_drvdata(pdev);
+
+	clk_disable_unprepare(st->pl);
+}
+
+static struct platform_driver fclk_driver = {
+	.driver = {
+		.name = KBUILD_MODNAME,
+		.of_match_table = fclk_of_match,
+	},
+	.probe		= fclk_probe,
+	.remove		= fclk_remove,
+};
+
+module_platform_driver(fclk_driver);
+
+MODULE_AUTHOR("Shubhrajyoti Datta <shubhrajyoti.datta@xilinx.com>");
+MODULE_DESCRIPTION("fclk enable");
+MODULE_LICENSE("GPL");
